pkg/util/recentcnt.go

```diff
@@ -0,0 +1,40 @@
+package util
+
+import "time"
+
+// RecentCounter is a struct that keeps track of recent timestamps within a given interval.
+type RecentCounter struct {
+	interval   time.Duration
+	timestamps []time.Time
+}
+
+func NewRecentCounter(interval time.Duration) *RecentCounter {
+	return &RecentCounter{
+		interval:   interval,
+		timestamps: make([]time.Time, 0),
+	}
+}
+
+func (rc *RecentCounter) cleanup(t time.Time) {
+	checkpoint := t.Add(-rc.interval)
+	i := 0
+	for ; i < len(rc.timestamps); i++ {
+		if rc.timestamps[i].After(checkpoint) {
+			break
+		}
+	}
+	rc.timestamps = rc.timestamps[i:]
+}
+
+// Inc increments the counter and adds the current timestamp to the list of timestamps.
+func (rc *RecentCounter) Inc() {
```

Comment:
Would need to check the time is non-decreasing.

Also, can do 
`
func (rc *RecentCounter) Inc() {
    rc.inc(time.Now())
}
`

and test the private interface.

Reply:
> Would need to check the time is non-decreasing

IIRC (not 100% sure) Go guarantees this, even when the system clock is not monotonic.

Reply:
> IIRC (not 100% sure) Go guarantees this

It won't be guaranteed if time is accepted through an argument.

Reply:
> IIRC (not 100% sure) Go guarantees this, even when the system clock is not monotonic.

Go does provide monotonic clocks. See the [time package docs](https://pkg.go.dev/time#hdr-Monotonic_Clocks).

> It may make sense to have Inc() and Get() also take the current time as input. More testable imo, and would have consistency with cleanup()

With this design, it's caller's responsibility to provide monotonic timestamps to `Inc` and `Get` functions. I don't think this is a good idea, since it's fragile and easy to break.

@Omrigan's idea is nice, however, looks a bit hacky.

Reply:
Addressed here 20a66f353704170fae0f1aa760905f4ce3674a1e.